catalog/app/containers/Search/model.ts

```diff
@@ -698,50 +706,328 @@ export function useNextPagePackagesQuery(after: string) {
   return folded
 }
 
-interface MetaFacetsQueryProps {
-  searchString: SearchUrlState['searchString']
-  buckets: SearchUrlState['buckets']
-  filter: PackagesFilterState
+export type NextPageQueryResult =
+  | ReturnType<typeof useNextPagePackagesQuery>
+  | ReturnType<typeof useNextPageObjectsQuery>
+
+interface NextPageQueryProps {
+  after: string
+  children: RenderFn<NextPageQueryResult>
+}
+
+export function NextPageObjectsQuery({ after, children }: NextPageQueryProps) {
+  return children(useNextPageObjectsQuery(after))
 }
 
-function useMetaFacetsQuery({ searchString, buckets, filter }: MetaFacetsQueryProps) {
-  const gqlFilter = PackagesSearchFilterIO.toGQL(filter)
-  return GQL.useQuery(META_FACETS_QUERY, { searchString, buckets, filter: gqlFilter })
+export function NextPagePackagesQuery({ after, children }: NextPageQueryProps) {
+  return children(useNextPagePackagesQuery(after))
 }
 
 const NO_FACETS: PackageUserMetaFacet[] = []
 
-export function usePackagesMetaFilters() {
-  const model = useSearchUIModel()
-  invariant(model.state.resultType === ResultType.QuiltPackage, 'Filter type mismatch')
+export const FacetsFilteringState = tagged.create(
+  'app/containers/Search:FacetsFilteringState' as const,
+  {
+    Disabled: () => {},
+    Enabled: (x: {
+      value: string
+      set: (value: string) => void
+      isFiltered: boolean
+      serverSide: boolean
+    }) => x,
+  },
+)
+
+export type FacetsFilteringStateInstance = tagged.InstanceOf<typeof FacetsFilteringState>
+
+export const AvailableFiltersState = tagged.create(
+  'app/containers/Search:AvailableFiltersState' as const,
+  {
+    Loading: () => {},
+    Empty: () => {},
+    Ready: (x: {
+      filtering: FacetsFilteringStateInstance
+      facets: {
+        available: readonly PackageUserMetaFacet[]
+        visible: FacetTree
+        hidden: FacetTree
+      }
+      fetching: boolean
+    }) => x,
+  },
+)
+
+export type AvailableFiltersStateInstance = tagged.InstanceOf<
+  typeof AvailableFiltersState
+>
+
+type RenderFn<T> = (arg: T) => JSX.Element | null
+
+interface RenderProps<T> {
+  children: RenderFn<T>
+}
+
+export function AvailablePackagesMetaFilters({
```

Comment:
I definitely can't help with this, because I don't fully follow the design and code paths, but as a general rule I try to use is to stay away from hooks outside UI. So, I would like to use hooks in UI code, and to memoize some final outputs from IO (graphql, https etc.) and from model, but keep a model straight javascript/typescript.

Reply:
the point is i need react logic here to memoize intermediate results, and i also need conditions, so neither i can use "pure" js (bc i would have to implement or borrow some stateful logic like streams or smth), nor hooks (bc no conditional support). some sufficiently advanced non-react solution would be nice (like, fc, effect-ts), but seems like too much trouble getting that in at that stage of development.